pkg/build/controller/controller.go

```diff
@@ -51,39 +58,77 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) {
+		build.Message = err.Error()
+	}
 
+	if _, err := bc.BuildUpdater.UpdateBuild(build.Namespace, build); err != nil {
+		glog.V(2).Infof("Failed to record changes to build %s/%s: %#v", build.Namespace, build.Name, err)
+	}
+}
+
+// nextBuildStatus updates build with any appropriate changes, or returns an error if
+// the change cannot occur. When returning nil, be sure to set build.Status and optionally
+// build.Message.
+func (bc *BuildController) nextBuildStatus(build *buildapi.Build) error {
 	// If a cancelling event was triggered for the build, update build status.
 	if build.Cancelled {
-		glog.V(2).Infof("Cancelling build %s.", build.Name)
-		nextStatus = buildapi.BuildStatusCancelled
-
-	} else {
-
-		var podSpec *kapi.Pod
-		var err error
-		if podSpec, err = bc.BuildStrategy.CreateBuildPod(build); err != nil {
-			glog.V(2).Infof("Strategy failed to create build pod definition: %v", err)
-			nextStatus = buildapi.BuildStatusFailed
-		} else {
-
-			if _, err := bc.PodManager.CreatePod(build.Namespace, podSpec); err != nil {
-				if !errors.IsAlreadyExists(err) {
-					glog.V(2).Infof("Failed to create pod for build %s: %#v", build.Name, err)
-					nextStatus = buildapi.BuildStatusFailed
-				}
-			} else {
-				glog.V(2).Infof("Created build pod: %#v", podSpec)
-				nextStatus = buildapi.BuildStatusPending
+		glog.V(4).Infof("Cancelling build %s.", build.Name)
+		build.Status = buildapi.BuildStatusCancelled
+		return nil
+	}
+
+	// lookup the destination from the referenced image repository
+	spec := build.Parameters.Output.DockerImageReference
+	if ref := build.Parameters.Output.To; ref != nil {
+		// TODO: security, ensure that the reference image stream is actually visible
+		namespace := ref.Namespace
+		if len(namespace) == 0 {
+			namespace = build.Namespace
+		}
+
+		repo, err := bc.ImageRepositoryClient.GetImageRepository(namespace, ref.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				build.Status = buildapi.BuildStatusFailed
+				build.Message = fmt.Sprintf("The referenced output image repository %s/%s does not exist", namespace, ref.Name)
+				return nil
 			}
+			return fmt.Errorf("the referenced output repo %s/%s could not be found by %s/%s: %v", namespace, ref.Name, build.Namespace, build.Name, err)
 		}
+		spec = repo.Status.DockerImageRepository
 	}
 
-	build.Status = nextStatus
-	if _, err := bc.BuildUpdater.UpdateBuild(build.Namespace, build); err != nil {
-		glog.V(2).Infof("Failed to update build %s: %#v", build.Name, err)
+	// set the expected build parameters, which will be saved if no error occurs
+	build.Status = buildapi.BuildStatusPending
+	build.PodName = fmt.Sprintf("build-%s", build.Name)
+
+	// override DockerImageReference in the strategy for the copy we send to the server
+	copy, err := kapi.Scheme.Copy(build)
+	if err != nil {
+		return fmt.Errorf("unable to copy build: %v", err)
+	}
+	buildCopy := copy.(*buildapi.Build)
+	buildCopy.Parameters.Output.DockerImageReference = spec
```

Comment:
I think this answers my earlier question about how DockerImageReference was getting populated in the build object.